numbast/src/numbast/static/struct.py

```diff
@@ -0,0 +1,691 @@
+
+        self.Imports.add("from numba.cuda.cudadecl import register")
+        self.Imports.add("from numba.cuda.cudadecl import register_global")
+        self.Imports.add("from numba.core.typing.templates import ConcreteTemplate")
+        self.Imports.add("from numba.types import Function")
+
+        signatures_str = ", ".join(signature_strs)
+
+        self._struct_ctor_typing_rendered = (
+            self.struct_ctor_template_typing_template.format(
+                struct_name=self._struct_name, signatures=signatures_str
+            )
+        )
+
+    def _render(self):
+        """Render all struct constructors."""
+
+        signatures: list[str] = []
+        for ctor_decl in self._ctor_decls:
+            renderer = StaticStructCtorRenderer(
+                struct_name=self._struct_name,
+                struct_type_class=self._struct_type_class,
+                struct_type_name=self._struct_type_name,
+                ctor_decl=ctor_decl,
+            )
+            renderer._render()
+
+            self._python_rendered += renderer._python_rendered
+            self._c_rendered += renderer._c_rendered
+
+            signatures.append(renderer.signature_str)
+
+        self._render_typing(signatures)
+
+        self._python_rendered += self._struct_ctor_typing_rendered
+
+    @property
+    def python_rendered(self) -> str:
+        """The python script that contains the bindings to all constructors."""
+        return self._python_rendered
+
+    @property
+    def c_rendered(self) -> str:
+        """The C program that contains the shim functions to all constructors."""
+        return self._c_rendered
+
+
+class StaticStructRenderer(BaseRenderer):
+    """Renderer that renders bindings to a single CUDA C++ struct.
+
+    Parameters
+    ----------
+    decl: Struct
+        Declaration of the struct.
+    parent_type: type
+        The parent Numba type of the new struct type created.
+    data_model: type
+        The data model of the new struct type.
+    header_path: os.PathLike
+        Path to the header that contains the declaration of the struct.
+    aliases: list[str], optional
+        TODO: If the struct has other aliases, specify them here. Numbast creates
+        aliased objects that references the original python API object.
+    """
+
+    typing_template = """
+# Typing for {struct_name}
+class {struct_name}_type_class({parent_type}):
+    def __init__(self):
+        super().__init__(name="{struct_name}")
+        self.alignof_ = {struct_alignof}
+        self.bitwidth = {struct_sizeof} * 8
+
+{struct_name}_type = {struct_name}_type_class()
+"""
+
+    python_api_template = """
+# Make Python API for struct
+{struct_name} = type("{struct_name}", (), {{"_nbtype": {struct_name}_type}})
+"""
+
+    primitive_data_model_template = """
+@register_model({struct_name}_type_class)
+class {struct_name}_model(PrimitiveModel):
+    def __init__(self, dmm, fe_type):
+        be_type = ir.IntType(fe_type.bitwidth)
```

Comment:
I ended up with an `int32` for a `half` data type in the source (using the myfloat16 example from the readme) - is that expected?

Reply:
Hmm, probably no? It should be `int16` no?

Reply:
I checked the `myfloat16` example above (less the double underscore prefix) and I saw that the generated numba type is below:
```python
class myfloat16_type_class(Number):
    def __init__(self):
        super().__init__(name="myfloat16")
        self.alignof_ = 2
        self.bitwidth = 2 * 8
```

This means the IR type used for this type is `int16`, not `int32`. Btw, additional note to the `myfloat8` example, it will also use 16bit because of `aligned(2)` in declaration. In this case, each `myfloat8` will still take 16 bits, so its numba type also have `.bitwidth` set to 16.